src/python/pants/engine/internals/mapper.py

```diff
@@ -32,37 +32,27 @@ class AddressMap:
     def parse(
         cls,
         filepath: str,
-        filecontent: bytes,
+        build_file_content: str,
         parser: Parser,
         extra_symbols: BuildFilePreludeSymbols,
     ) -> "AddressMap":
         """Parses a source for addressable Serializable objects.
 
-        No matter the parser used, the parsed and mapped addressable objects are all 'thin'; ie: any
-        objects they point to in other namespaces or even in the same namespace but from a separate
-        source are left as unresolved pointers.
-
-        :param filepath: The path to the byte source containing serialized objects.
-        :param filecontent: The content of byte source containing serialized objects to be parsed.
-        :param parser: The parser cls to use.
+        The parsed and mapped addressable objects are all 'thin': any objects they point to in other
+        namespaces or even in the same namespace but from a separate source are left as unresolved
+        pointers.
         """
         try:
-            objects = parser.parse(filepath, filecontent, extra_symbols)
+            objects = parser.parse(filepath, build_file_content, extra_symbols)
         except Exception as e:
             raise MappingError(f"Failed to parse {filepath}:\n{e!r}")
         objects_by_name: Dict[str, ThinAddressableObject] = {}
         for obj in objects:
-            if not Serializable.is_serializable(obj):
-                raise UnaddressableObjectError("Parsed a non-serializable object: {!r}".format(obj))
             attributes = obj._asdict()
-
-            name = attributes.get("name")
-            if not name:
-                raise UnaddressableObjectError("Parsed a non-addressable object: {!r}".format(obj))
```

Comment:
The Callback parser will never return objects that are not serializable or don't have a name. I spent some time both testing this out in the wild, along with reasoning about it. The only Struct we expect to have is a TargetAdaptor, which by definition is Serializable, and has special code to ensure `name` is always set.